proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,313 @@
+cluster restart, it is used to verify received shreds and filter Gossip peers. 
+In the proposed optimistic cluster restart plan, we introduce a new temporary
+shred version in the silent repair phase so validators in restart don't 
+interfere with those not in restart. Currently this silent repair shred version 
+is calculated using `(current_shred_version + 1) % 0xffff`.
+
+* `RESTART_STAKE_THRESHOLD`: We need enough validators to participate in a
+restart so they can make decision for the whole cluster. If everything works
+perfect, we only need 2/3 of the total stake. However, validators could die
+or perform abnormally, so we currently set the `RESTART_STAKE_THRESHOLD` at
+80%, which is the same as now.
+
+## Motivation
+
+Currently during a cluster restart, validator operators need to decide the
+highest optimistically confirmed slot, then restart the validators with new
+command-line arguments.
+
+The current process involves a lot of human intervention, if people make a
+mistake in deciding the highest optimistically confirmed slot, it is
+detrimental to the viability of the ecosystem.
+
+We aim to automate the negotiation of highest optimistically confirmed slot and
+the distribution of all blocks on that fork, so that we can lower the 
+possibility of human mistakes in the cluster restart process. This also reduces 
+the burden on validator operators, because they don't have to stay around while 
+the validators automatically try to reach consensus, the validator will halt
+and print debug information if anything goes wrong, and operators can set up
+their own monitoring accordingly.
+
+## Alternatives Considered
+
+### Automatically detect outage and perform cluster restart
+
+The reaction time of a human in case of emergency is measured in minutes,
+while a cluster restart where human initiate validator restarts takes hours.
+We considered various approaches to automatically detect outage and perform
+cluster restart, which can reduce recovery speed to minutes or even seconds.
+
+However, automatically restarting the whole cluster seems risky. Because
+if the recovery process itself doesn't work, it might be some time before
+we can get human's attention. And it doesn't solve the cases where new binary
+is needed. So for now we still plan to have human in the loop.
+
+After we gain more experience with the restart approach in this proposal, we
+may slowly try to automate more parts to improve cluster reliability.
+
+### Use gossip and consensus to figure out restart slot before the restart
+
+The main difference between this and the current proposal is this alternative
+tries to make the cluster automatically enter restart preparation phase without 
+human intervention.
+
+While getting human out of the loop improves recovery speed, there are concerns 
+about recovery gossip messages interfering with normal gossip messages, and 
+automatically start a new message in gossip seems risky.
```

Comment:
This paragraph seems unnecessary as it doesn't really contribute to the protocol spec.

Reply:
It is mainly discussing why we didn't choose this alternative.